p2p/p2p.go

```diff
@@ -188,3 +201,87 @@ func (s *Service) ListenAddrs() ([]multiaddr.Multiaddr, error) {
 
 	return listenAddrs, nil
 }
+
+// NewStream creates a bidirectional connection to a random peer that implements a set of protocol ids
+func (s *Service) NewStream(ctx context.Context, pids ...protocol.ID) (network.Stream, error) {
+	peers := s.host.Peerstore().Peers()
+	peersCount := peers.Len()
+	if peersCount < 0 {
+		return nil, errors.New("no peers")
+	}
+
+	randomPeerIdx := rand.Intn(peersCount) //nolint: gosec
+	for peerIdx := (randomPeerIdx + 1) % peersCount; ; peerIdx = (peerIdx + 1) % peersCount {
```

Comment:
What is the end condition here? It will loop forever? Also, you might wanna keep a list of peer with known protocol. I'm not sure how libp2p handle list of peer in its peer store, it could store a lot of peer that does not support the protocol. 

See https://github.com/NethermindEth/juno/pull/815/files#diff-cce63aa475cb9a161db5de4cf7297f3ef5fc058000702e26e6dbcd8c195ac9daR77.

Reply:
Might wanna have a wait and cancellation on context cancelled.

Reply:
L223-225 is the end condition. It gives up, after looping once thru the entire peerlist

Reply:
> I'm not sure how libp2p handle list of peer in its peer store, it could store a lot of peer that does not support the protocol

Yeap, we can deal with advanced peer management later.

Reply:
> Might wanna have a wait and cancellation on context cancelled.

Added explicit checks for them, thanks!


https://github.com/NethermindEth/juno/compare/2ca888f3df1a46114e759f9f6274d7cf50550b9e..6c5c1189572e37f385a713e73db46652e7cf61a8